tiledb/api/c_api/group/group_api.cc

```diff
@@ -0,0 +1,797 @@
+  if (*uri == nullptr) {
+    return TILEDB_ERR;
+  }
+  std::memcpy(*uri, uri_str.value().data(), uri_str.value().size());
+  (*uri)[uri_str.value().size()] = '\0';
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_group_get_is_relative_uri_by_name(
+    tiledb_group_handle_t* group,
+    const char* name,
+    uint8_t* is_relative) {
+  ensure_group_is_valid(group);
+  if (name == nullptr) {
+    throw CAPIStatusException("argument `name` may not be nullptr");
+  }
+  ensure_output_pointer_is_valid(is_relative);
+
+  auto&& [st, uri_str, object_type, name_str, relative] =
+      group->group().member_by_name(name);
+  throw_if_not_ok(st);
+
+  *is_relative = *relative ? 1 : 0;
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_group_is_open(
+    tiledb_group_handle_t* group, int32_t* is_open) {
+  ensure_group_is_valid(group);
+  ensure_output_pointer_is_valid(is_open);
+
+  *is_open = static_cast<int32_t>(group->group().is_open());
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_group_get_uri(
+    tiledb_group_handle_t* group, const char** group_uri) {
+  ensure_group_is_valid(group);
+  ensure_output_pointer_is_valid(group_uri);
+
+  // PJD: Is this awkward that the other string returning
+  // functions allocated storage, but this one does not?
+  *group_uri = group->group().group_uri().c_str();
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_group_get_query_type(
+    tiledb_group_handle_t* group, tiledb_query_type_t* query_type) {
+  ensure_group_is_valid(group);
+  ensure_output_pointer_is_valid(query_type);
+
+  // Get query_type
+  tiledb::sm::QueryType type;
+  throw_if_not_ok(group->group().get_query_type(&type));
+
+  *query_type = static_cast<tiledb_query_type_t>(type);
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_group_dump_str(
+    tiledb_group_handle_t* group,
+    char** dump_ascii,
+    const uint8_t recursive) {
+  ensure_group_is_valid(group);
+  ensure_output_pointer_is_valid(dump_ascii);
+
+  const std::string str = group->group().dump(2, 0, recursive);
+
+  *dump_ascii = static_cast<char*>(std::malloc(str.size() + 1));
+  if (*dump_ascii == nullptr) {
+    return TILEDB_ERR;
+  }
+
+  std::memcpy(*dump_ascii, str.data(), str.size());
+  (*dump_ascii)[str.size()] = '\0';
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_serialize_group(
+    tiledb_ctx_handle_t* ctx,
+    const tiledb_group_handle_t* group,
+    tiledb_serialization_type_t serialize_type,
+    int32_t client_side,
+    tiledb_buffer_t** buffer) {
+  ensure_context_is_valid(ctx);
+  ensure_group_is_valid(group);
+
+  // client_side is currently unused
+  (void)client_side;
+
+  // Allocate a buffer list
+  // PJD: This seems suboptimal in error reporting. Perhaps add
+  // an OOM return? Then again the fancy way may involve just
+  // wrapping tiledb_buffer_t with a handle?
+  if (tiledb_buffer_alloc(ctx, buffer) != TILEDB_OK || sanity_check(ctx, *buffer) == TILEDB_ERR)
+    return TILEDB_ERR;
```

Comment:
This code that returns if the buffer can't be allocated is repeated (and will be used elsewhere). This is a good time to write a utility function that (1) returns a buffer handle or (2) throws if it can't create a handle to return. Call it `buffer_make_handle` by analogy with `tiledb_buffer_handle_t::make_handle`.

Reply:
Fixed by properly wrapping `tiledb_buffer_t`